service/handlers.go

```diff
@@ -1210,20 +1219,55 @@ func (h handler) listTokens(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 
-	tokens, err := h.dbClient.ListTokenEntries(ctx, projectName)
+	dbTokens, err := h.dbClient.ListTokenEntries(ctx, projectName)
+	if err != nil {
+		level.Error(l).Log("message", "error listing project tokens", "error", err)
+		h.errorResponse(w, "error listing project tokens", http.StatusInternalServerError)
+		return
+	}
+
+	cpTokens, err := cp.ListProjectTokens(projectName)
 	if err != nil {
 		level.Error(l).Log("message", "error listing project tokens", "error", err)
 		h.errorResponse(w, "error listing project tokens", http.StatusInternalServerError)
 		return
 	}
 
 	resp := []responses.ListTokens{}
-	for _, tokenEntry := range tokens {
-		resp = append(resp, responses.ListTokens{
-			CreatedAt: tokenEntry.CreatedAt,
-			ExpiresAt: tokenEntry.ExpiresAt,
-			TokenID:   tokenEntry.TokenID,
-		})
+
+	// Vault automatically deletes secrets when they expire, so there is a chance that the tokens DB and Vault
+	// are not in sync.
+	//
+	// Check if the two project tokens list are equal. If not equal, return CP tokens as the CP
+	// is the source of truth.
+	// Else, return the DB tokens as they already contain the expire time.
+	//
+	// This is assuming that the expire time for the DB token is correct since the token exists in the CP.
+	if isProjectTokensListsEqual(cpTokens, dbTokens) {
+		for _, tokenEntry := range dbTokens {
+			resp = append(resp, responses.ListTokens{
+				CreatedAt: tokenEntry.CreatedAt,
+				ExpiresAt: tokenEntry.ExpiresAt,
+				TokenID:   tokenEntry.TokenID,
+			})
+		}
+	} else {
+		for _, projectToken := range cpTokens {
+			// cp.ListProjectTokens only returns the token ID. Call GetProjectToken
+			// to populate the necessary fields.
+			getProjectTkn, err := cp.GetProjectToken(projectName, projectToken.ID)
```

Comment:
This can be removed if `cp.ListProjectTokens` can return the expire, creation time, etc. It can be done, but the Vault unit test will need to be refactored to use the generated creds provider mocks. What do you guys think?

Reply:
i think it's a good move if `cp.ListProjectTokens()` can just return everything for us.